File: javascript/graph/graph.js
'use strict';

let Queue = require('../queues/queues.js');
let Stack = require('../stacks/stacks.js');

class Vertex {
  constructor(value) {
    this.value = value;
  }
}

class Edge {
  constructor(vertex, weight = 0) {
    this.vertex = vertex;
    this.weight = weight;
  }
}

class Graph {
  constructor() {
    this.adjacencyList = new Map();
  }


  addVertex(vertex) {
    this.adjacencyList.set(vertex, []);
    return this;
  }


  addDirectedEdge(startVertex, endVertex) {
    if (this.adjacencyList.has(startVertex) && this.adjacencyList.has(endVertex)) {
      const neighbors = this.adjacencyList.get(startVertex);
      neighbors.push(new Edge(endVertex));
      return this;
    } else {
      console.error('Vertex may not be present in this graph');
      return null;
    }
  }


  addUndirectedEdge(startVertex, endVertex) {
    this.addDirectedEdge(startVertex, endVertex);
    this.addDirectedEdge(endVertex, startVertex);
    return this;
  }


  getSize() {
    return this.adjacencyList.size;
  }


  getNeighbors(vertex) {
    return this.adjacencyList.get(vertex);
  }


  bft(startVertex) {

    const queue = new Queue();
    const visitedNodes = new Set();
    queue.enqueue(startVertex);
    visitedNodes.add(startVertex);

    while (queue.peek()) {

      const currentVertex = queue.dequeue();
      const neighbors = this.adjacencyList.get(currentVertex);

      for (let neighbor of neighbors) {
        const neighborVertex = neighbor.vertex;
        if (visitedNodes.has(neighborVertex)) {
          continue;
        } else {
          visitedNodes.add(neighborVertex);
        }
        queue.enqueue(neighborVertex);
      }
    }
    return visitedNodes;
  }


  dft(startVertex) {

    const stack = new Stack();
    const visitedNodes = new Set();
    stack.push(startVertex);
    visitedNodes.add(startVertex);

    while(stack.peek()) {

      let currentVertex = stack.pop();
      let neighborList = this.adjacencyList.get(currentVertex);

      for(let edge of neighborList) {
        let neighborVertex = edge.vertex;
        if(visitedNodes.has(neighborVertex)) {
          continue;
        } else {
          visitedNodes.add(neighborVertex);
        }
        stack.push(neighborVertex);
      }
    }
    return visitedNodes;
  }

}


module.exports = {Vertex, Edge, Graph};

File: python/graph/graph.py
from queues.queues import Queue


class Graph:
    def __init__(self):
        self._adjacency_list = {}

    def add_node(self, node):
        """
        add node
        Arguments: value
        Returns: The added node
        Add a node to the graph
        """
        # v = Vertex(value)
        self._adjacency_list[node] = []
        return node

    def add_edge(self, start_vertex, end_vertex, weight=0):
        """
        add edge
        Arguments: 2 nodes to be connected by the edge, weight (optional)
        Returns: nothing
        Adds a new edge between two nodes in the graph
        If specified, assign a weight to the edge
        Both nodes should already be in the Graph
        """
        if start_vertex not in self._adjacency_list:
            raise KeyError("Starting Vertex not in Graph")
        if end_vertex not in self._adjacency_list:
            raise KeyError("End Vertex not in Graph")

        edge = Edge(end_vertex, weight)
        adjacencies = self._adjacency_list[start_vertex]
        adjacencies.append(edge)

    def get_nodes(self):
        """
        get nodes
        Arguments: none
        Returns all of the nodes in the graph as a collection (set, list, or similar)
        """
        return self._adjacency_list.keys()

    def get_neighbors(self, vertex):
        """
        get neighbors
        Arguments: node
        Returns a collection of edges connected to the given node
        Include the weight of the connection in the returned collection
        """
        return self._adjacency_list[vertex]

    def size(self):
        """
        size
        Arguments: none
        Returns the total number of nodes in the graph
        """
        return len(self._adjacency_list)

    def breadth_first(self, vertex):
        nodes = []
        breadth = Queue()
        visited = set()

        breadth.enqueue(vertex)
        visited.add(vertex)

        while not breadth.isEmpty():
            front = breadth.dequeue()
            nodes.append(front.value)

            for child in [e.vertex for e in self.get_neighbors(front)]:
                if child not in visited:
                    visited.add(child)
                    breadth.enqueue(child)

        return nodes


class Vertex:
    def __init__(self, value=None):
        self.value = value


class Edge:
    def __init__(self, vertex, weight=0):
        self.vertex = vertex
        self.weight = weight
